Initial set of RFCs for Az PowerShell (#10045)
Azure/azure-powershell: documentation/RFC/RFC0002-SubscriptionList-in-Get.md

---
RFC: RFC0002
Author: Azure PowerShell team
Status:  Draft
SupercededBy: n/a
Version: 1.0
---

# List of subscriptions for GET commands

A GET cmdlet is mostly used to obtain an Azure resource or validate its existence in Azure.
With more and more customers using multiple subscriptions the ability to search for resources accross subscriptions has become a more needed convenience.

> We are planning to run GET accross  

Reviewer: I hope this was intended to be finished with "across all subscriptions within a single Azure Tenant"

As a consultant in many tenants I don't think this functionality should be able to reach across tenants but would be extremely useful within a single tenant

Author: Thanks @kilasuit, I missed to finish this sentence.

## Motivation

Explain the benefits of the change for the users.

```code
As a PowerShell script developer
I can run a GET command accross several subscriptions.

> so that I do not have to maintain a list of my suscriptions

Reviewer 2: ```suggestion
so that I do not have to switch contexts to access and manage resources in multiple subscriptions
```

> ```

Reviewer 2: ```
As a powershell script developer
I can run a GET command across multiple subscriptions
So I can search for Azure resources that meet specific criteria no matter what subscription they are in.

Author: @markcowl this is much better!

## User Experience

> The default behavior will search accross all the subscriptions you have access:

Reviewer 2: ```suggestion
The default behavior will not change, but the user can specify a different subscription, or multiple subscriptions when performing gats.  The output of these gets can be filtered and piped into any other cmdlet.
```

Reviewer: @markcowl - the default behaviour should ideally be a search across all available subscriptions unless filtered to current subscription.

However a potential middle ground would be to add a switch parameter `-AllSubscriptions` that allows the search to run across all subscriptions.

Reviewer 2: Yes, this is currently about what preview 1 will do.  We made minimal changes in Az.Accounts for this preview - we will be able to do more in preview 2

```PowerShell
Get-AzVM -name "bla"
```

One can specify one of several subscriptions to use:

```PowerShell

> Get-AzVm -name "bla" -SubscriptionId xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx

Reviewer 2: ```suggestion
Get-AzVm -name "bla" -SubscriptionId xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx | Restart-AzVM
```

> Get-AzVm -name "bla" -SubscriptionId ['xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx']

Reviewer 2: ```suggestion
Get-AzVm -name "bla" -SubscriptionId ['xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx', 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'] | Restart-AzVM
```

```

## Comments and Questions

Please insert below your anwser to the following questions or add any additional questions that you may have:

> - Should the default behavior look after all accessible subscriptions?

Reviewer 2: ```suggestion
- Should the default behavior select all subscriptions in a tenant as part of connecting an account?
```

  - _Your prefered default behavior_

> - Should the default behavior be only looking at the default subscriptions?

Reviewer 2: ```suggestion
- Should we provide a cmdlet to select a subset of subscriptions as the default, or would you prefer to use `$PSDefaultParameterValues` to do this?
```

Author: Thanks @markcowl I have slightly modified this one.
